src/irmin-pack/unix/io.ml

```diff
@@ -0,0 +1,282 @@
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open! Import
+open Io_intf
+
+(* File utils, taken from index.unix package *)
+module Util = struct
+  module Syscalls = Index_unix.Syscalls
+
+  (* NOTE following copied from Index_unix.Raw *)
+
+  let really_write fd fd_offset buffer buffer_offset length =
+    let rec aux fd_offset buffer_offset length =
+      let w = Syscalls.pwrite ~fd ~fd_offset ~buffer ~buffer_offset ~length in
+      if w = 0 || w = length then ()
+      else
+        (aux [@tailcall])
+          (fd_offset ++ Int63.of_int w)
+          (buffer_offset + w) (length - w)
+    in
+    aux fd_offset buffer_offset length
+
+  let really_read fd fd_offset length buffer =
+    let rec aux fd_offset buffer_offset length =
+      let r = Syscalls.pread ~fd ~fd_offset ~buffer ~buffer_offset ~length in
+      if r = 0 then buffer_offset (* end of file *)
+      else if r = length then buffer_offset + r
+      else
+        (aux [@tailcall])
+          (fd_offset ++ Int63.of_int r)
+          (buffer_offset + r) (length - r)
+    in
+    aux fd_offset 0 length
+end
+
+module type S = S
+
+module Unix = struct
+  type misc_error = Unix.error * string * string
+  type create_error = [ `Io_misc of misc_error ]
+
+  type open_error =
+    [ `Io_misc of misc_error | `No_such_file_or_directory | `Is_a_directory ]
+
+  type read_error =
+    [ `Io_misc of misc_error | `Read_out_of_bounds | `Read_on_closed ]
+
+  type write_error =
+    [ `Io_misc of misc_error | `Ro_not_allowed | `Write_on_closed ]
+
+  type close_error = [ `Io_misc of misc_error | `Double_close ]
+
+  type mkdir_error =
+    [ `Io_misc of misc_error | `File_exists | `No_such_file_or_directory ]
+
+  type move_file_error = [ `Io_misc of misc_error ]
+
+  (* the interface indicates that we need to catch double close errors; so type t needs
+     some field to record whether the instance has been closed; we also need a field for
+     "readonly" (to avoid querying the fd for this info); also for path (because of "path"
+     function below) *)
+  type t = {
+    fd : Unix.file_descr;
+    mutable closed : bool;
+    readonly : bool;
+    path : string;
+  }
+
+  exception Read_error of read_error
+  exception Write_error of write_error
+
+  (* default permissions which are required by some functions *)
+
+  let default_create_perm = 0o660
+  let default_open_perm = 0o660
+  let default_mkdir_perm = 0o770 (* or 0o777? *)
+
+  (* FIXME meaning of overwrite? we assume it means "overwrite: if the file exists,
+     truncate it and use the truncated file instead"; otherwise, if overwrite=false, and
+     we try to create a file that already exists, we fail with EEXIST *)
+  let create ~path ~overwrite : (t, [> create_error ]) result =
+    try
+      (* NOTE there is a race condition between checking for file existence, and acting on
```

Comment:
maybe add in the comment that only readwrite instances call this function, and that we only guarantee single writer.

Reply:
Yes, we don't have to care about the race conditions of rw operations